File: Java/Merge k Sorted Lists.java
class Merge k Sorted Lists{
	public ListNode mergeKLists(ArrayList<ListNode> lists) {
       if(lists == null || lists.size() == 0)
            return null;
        PriorityQueue<ListNode> minHeap = new PriorityQueue<ListNode>(lists.size(), new Comparator<ListNode>(){
        	public int compare(ListNode n1,ListNode n2){
        		return n1.val - n2.val;
        	}
        });
        for(ListNode node:lists){
        	if(node != null)
        		minHeap.offer(node);
        }
        ListNode dummy = new ListNode(-1);
        ListNode p = dummy;
        while(!minHeap.isEmpty()){
        	ListNode temp = minHeap.poll();
        	if(temp.next != null)
        		minHeap.offer(temp.next);
        	p.next = temp;
        	p = p.next;
        }
        return dummy.next;
    }
}

File: Python/Merge k Sorted Lists.py
# Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.


# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
 
class Solution:
    # @param a list of ListNode
    # @return a ListNode
    def mergeKLists(self, lists):
        heap = []
        for node in lists:
            if node != None: heap.append((node.val, node))
        heapq.heapify(heap)
        head = ListNode(0); curr = head
        while heap:
            pop = heapq.heappop(heap)
            curr.next = ListNode(pop[0])
            curr = curr.next
            if pop[1].next: heapq.heappush(heap, (pop[1].next.val, pop[1].next))
        return head.next